keps/sig-node/20200327-not-inject-kubelet-to-its-components.md

```diff
@@ -0,0 +1,318 @@
+It makes kubelet root directory too large to extend and these components' dependency blurry.
+
+## Motivation
+
+Avoid to init component by its parent structure. We should make kubelet better modularized and slim down both `Kubelet` and its methods.
+
+### Pros
+- Make kubelet(very huge now) better Modularized by moving some members(hostname, nodeName, clock, provider, etc) to a new one called BasicInfo.
+- Slim down pkg/kubelet directory by returning some kubelet method to kubeGenericRuntimeManager, resourceAnalyzer.
+- Resolve some circular dependencies like kubeGenericRuntimeManager and volumeManager, StatsProvider and ResourceAnalyzer.
+
+Kubelet components dependency states and migration plan will be listed in Proposal.
+
+### Cons
+I think this migration with low risk, because functions won't be changed and this process can be protected by unit tests and e2e tests.
+
+### Goals
+- Do not inject kubelet to its components' initialization.
+- Slim down Kubelet structure and its methods.
+
+### Non-Goals
+- Any effect on kubelet current function.
+
+## Proposal
+
+### Migrate Kubelet basic members to BasicInfo
+
+Create `BasicInfo` locates in `pkg/kubelet/basicinfo` only contains basic members which often used by its components.
+
+```go
+// Not import other kubelet component.
+import (
+    "k8s.io/api/core/v1"
+    "k8s.io/apimachinery/pkg/apis/meta/v1"
+    "k8s.io/apimachinery/pkg/labels"
+    "k8s.io/apimachinery/pkg/types"
+    "k8s.io/apimachinery/pkg/util/clock"
+    "k8s.io/client-go/listers/core/v1"
+    "k8s.io/client-go/kubernetes"
+    "k8s.io/client-go/tools/record"
+    "k8s.io/cloud-provider"
+    "k8s.io/cloud-provider/api"
+    "k8s.io/klog"
+    "k8s.io/kubernetes/pkg/apis/core"
+    "k8s.io/kubernetes/pkg/apis/core/v1"
+    "k8s.io/kubernetes/pkg/kubelet/apis"
+    "k8s.io/kubernetes/pkg/kubelet/config"
+    "k8s.io/kubernetes/pkg/util/node"
+    "k8s.io/kubernetes/pkg/util/taints"
+    "k8s.io/kubernetes/pkg/volume/util"
+)
+
+// BasicInfo's members are migrated from Kubelet.
+type BasicInfo struct {
+
+    Hostname string
+
+    NodeName types.NodeName
+
+    Clock clock.Clock
+
+    RegisterNode bool
+
+    RegistrationCompleted bool
+
+    RootDirectory string
+
+    KubeClient clientset.Interface
+
+    MasterServiceNamespace string
+
+    nodeLabels map[string]string
+
+    Recorder record.EventRecorder
+
+    ServiceLister serviceLister
+
+    NodeLister corelisters.NodeLister
+
+    providerID string
+
+    externalCloudProvider bool
+
+    Cloud cloudprovider.Interface
+
+    registerSchedulable bool
+
+    registerWithTaints []api.Taint
+
+    enableControllerAttachDetach bool
+
+    ExperimentalHostUserNamespaceDefaulting bool
+
+    NodeStatusFuncs []func(*v1.Node) error
+
+    KeepTerminatedPodVolumes bool // DEPRECATED
+}
+
+// these functions are kubelet's method now.
+type Interface struct {
```

Comment:
This seems like a generally strange grouping of functions.  If we want to introduce new interfaces, we should scope each interface to a single, logical unit.  E.g. if we can group all of the node status update stuff together, then that might make one good interface.  But it is hard to see how GetVolumeDevicePluginDir is conceptually related to SetNodeStatus.

Reply:
yep, split BasicInfo and the methods into several scopes looks better.

Reply:
Done